h2o-automl/src/main/java/ai/h2o/automl/targetencoding/BroadcastJoinForTargetEncoder.java

```diff
@@ -0,0 +1,185 @@
+   * @param leftFoldColumnIdx index of the fold column from `leftFrame` or `-1` if we don't use folds
+   * @param broadcastedFrame supposedly small frame that we will broadcast to all nodes and use it as lookup table for joining
+   * @param rightCatColumnsIdxs indices of the categorical columns from `broadcastedFrame` we want to use to calculate encodings 
+   * @param rightFoldColumnIdx index of the fold column from `broadcastedFrame` or `-1` if we don't use folds
+   * @return
+   */
+  static Frame join(Frame leftFrame, int[] leftCatColumnsIdxs, int leftFoldColumnIdx, Frame broadcastedFrame, int[] rightCatColumnsIdxs, int rightFoldColumnIdx, int numberOfFolds) {
+    int numeratorIdx = broadcastedFrame.find(TargetEncoder.NUMERATOR_COL_NAME);
+    int denominatorIdx = broadcastedFrame.find(TargetEncoder.DENOMINATOR_COL_NAME);
+
+    int broadcastedFrameCatCardinality = broadcastedFrame.vec(rightCatColumnsIdxs[0]).cardinality();
+
+    // Note: for our goals there is no need to store it in 2d array
+    // Use `broadcastedFrame` as major frame(first argument) while computing level mappings as it contains all the levels from `training` dataset
+    int[][] levelMappings = {CategoricalWrappedVec.computeMap( broadcastedFrame.vec(0).domain(), leftFrame.vec(leftCatColumnsIdxs[0]).domain())};
+    
+    int[][] encodingDataMap = new FrameWithEncodingDataToArray(rightCatColumnsIdxs[0], rightFoldColumnIdx, numeratorIdx, denominatorIdx, broadcastedFrameCatCardinality, numberOfFolds)
+            .doAll(broadcastedFrame)
+            .getEncodingDataArray();
+    new BroadcastJoiner(leftCatColumnsIdxs, leftFoldColumnIdx, encodingDataMap, levelMappings).doAll(leftFrame);
+    return leftFrame;
+  }
+
+  static class BroadcastJoiner extends MRTask<BroadcastJoiner> {
+    int _categoricalColumnIdx, _foldColumnIdx;
+    int[][] _encodingDataArray;
+    int[][] _levelMappings;
+
+    BroadcastJoiner(int[] categoricalColumnsIdxs, int foldColumnIdx, int[][] encodingDataMap, int[][] levelMappings) {
+      assert categoricalColumnsIdxs.length == 1 : "Only single column target encoding(i.e. one categorical column is used to produce its encodings) is supported for now";
+
+      _categoricalColumnIdx = categoricalColumnsIdxs[0];
+      _foldColumnIdx = foldColumnIdx;
+      _encodingDataArray = encodingDataMap;
+      _levelMappings = levelMappings;
+    }
+
+    @Override
+    public void map(Chunk[] cs) {
+      Chunk categoricalChunk = cs[_categoricalColumnIdx];
+      int numOfVecs = cs.length;
+      Chunk num = cs[numOfVecs - 2];
+      Chunk den = cs[numOfVecs - 1];
+      for (int i = 0; i < num.len(); i++) {
+        int levelValue = (int) categoricalChunk.at8(i);
+        int mappedLevelValue = -1;
+        try {
+          mappedLevelValue = _levelMappings[0][levelValue];
+        } catch (Exception ex) {
+          Log.info("Categorical level with index `" + levelValue + "` was not present in training data. ");
```

Comment:
logging events on a row-level should be avoided, it will have an adverse effect on performance and might pollute logs and render them unusable